drivers/watchdog/wdt_sam4l.c

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (C) 2024-2025, Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam4l_watchdog
+
+/**
+ * @brief Watchdog (WDT) Driver for Atmel SAM4L MCUs
+ *
+ * Note:
+ * - SAM4L watchdog has a fuse bit to automatically enable the watchdog at boot.
+ *   It should be enabled to keep compatibility with SAM family.
+ * - Since the MCU boots with WDT enabled, the CONFIG_WDT_DISABLE_AT_BOOT
+ *   is set default at boot and watchdog module is disabled in the MCU for
+ *   systems that don't need watchdog functionality.
+ * - If the application needs to use the watchdog in the system, then
+ *   CONFIG_WDT_DISABLE_AT_BOOT must be unset in the app's config file
+ */
+
+#include <zephyr/drivers/watchdog.h>
+#include <zephyr/drivers/clock_control/atmel_sam_pmc.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_sam4l);
+
+#define WDT_FIRST_KEY		  0x55ul
+#define WDT_SECOND_KEY		  0xAAul
+
+enum wdt_sam4l_clock_src {
+	WDT_SAM4L_CLK_SRC_RCSYS   = 0,
+	WDT_SAM4L_CLK_SRC_OSC32K  = 1,
+};
+
+struct wdt_sam4l_dev_cfg {
+	Wdt *regs;
+	void (*irq_cfg_func)(const struct device *dev);
+	const struct atmel_sam_pmc_config clock_cfg;
+	enum wdt_sam4l_clock_src clock_src;
+	bool lock;
+};
+
+struct wdt_sam4l_dev_data {
+	wdt_callback_t cb;
+	uint32_t flags;
+};
+
+/**
+ * @brief Sets the Watchdog Timer Control register to the @a ctrl value.
+ *
+ * @param ctrl  Value to set the WatchDog Timer Control register to.
+ */
+static void wdt_sam4l_set_ctrl(const struct wdt_sam4l_dev_cfg *cfg,
+			       const uint32_t ctrl)
+{
+	Wdt *const wdt = cfg->regs;
+	volatile uint32_t delay;
+
+	/* Calculate delay for internal synchronization, see 45.1.3 WDT errata */
+	if (cfg->clock_src == WDT_SAM4L_CLK_SRC_OSC32K) {
+		delay = DIV_ROUND_UP(SOC_ATMEL_SAM_MCK_FREQ_HZ * 2, SOC_ATMEL_SAM_RC32K_NOMINAL_HZ);
+	} else {
+		delay = DIV_ROUND_UP(SOC_ATMEL_SAM_MCK_FREQ_HZ * 2, SOC_ATMEL_SAM_RCSYS_NOMINAL_HZ);
+	}
+	/* ~8 cycles for one while loop */
+	delay >>= 3;
+	while (delay--) {
+	}
+	wdt->CTRL = WDT_CTRL_KEY(WDT_FIRST_KEY)
+		  | ctrl;
+	wdt->CTRL = WDT_CTRL_KEY(WDT_SECOND_KEY)
+		  | ctrl;
+}
+
+/**
+ * @brief Calculate timeout scale factor based on a input in milliseconds
+ *
+ * timeout(ms) = (2pow(scale + 1) * 1000) / wdt_clk
+ *
+ * @param cfg  Configuration
+ * @param time Timeout value in milliseconds
+ */
+static int wdt_sam4l_calc_timeout(const struct wdt_sam4l_dev_cfg *cfg,
+				  uint32_t time)
+{
+	uint32_t wdt_clk = cfg->clock_src == WDT_SAM4L_CLK_SRC_OSC32K
+			 ? SOC_ATMEL_SAM_RC32K_NOMINAL_HZ
+			 : SOC_ATMEL_SAM_RCSYS_NOMINAL_HZ;
+
+	for (int scale = 7; scale <= 31; ++scale) {
+		uint32_t timeout = (BIT(scale + 1) * 1000ull) / wdt_clk;
```

Comment:
should this be < and not <= ? BIT(31 + 1) overflows, does it not?

Reply:
The calculus is correct.
4294967296×1000/32768 -> 131072000 (worst case)

But I should use BIT64() instead BIT()

Fixed!